src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php

```diff
@@ -252,6 +254,54 @@ private function validateAndDenormalize($currentClass, $attribute, $data, $forma
                 $data = [$data];
             }
 
+            // In XML and CSV all basic datatypes are represented as strings, it is e.g. not possible to determine,
+            // if a value is meant to be a string, float, int or a boolean value from the serialized representation.
+            // That's why we have to transform the values, if one of these non-string basic datatypes is epxected.
+            //
+            // This is special to xml and csv format
+            if (
+                \is_string($data) &&
+                (XmlEncoder::FORMAT === $format || CsvEncoder::FORMAT === $format) &&
+                \in_array($type->getBuiltinType(), [Type::BUILTIN_TYPE_BOOL, Type::BUILTIN_TYPE_INT, Type::BUILTIN_TYPE_FLOAT], true)
```

Comment:
This final condition is redundant since it's already checked in the switch case.